tests/test_runner.py

```diff
@@ -31,3 +42,39 @@ def test_tag_selection(testfolder: str) -> None:
     r.collect(PATH, tags=("tag2",))
     assert len(r.benchmarks) == 1
     r.clear()
+
+
+def test_context_collection_in_runner(testfolder: str) -> None:
+    r = BenchmarkRunner()
+
+    context_providers = [system, cpuarch, python_version]
+    result = r.run(
+        os.path.join(testfolder),
+        tags=("standard",),
+        params={"x": 1, "y": 1},
+        context=context_providers,
+    )
+
+    print(result)
+    assert "system" in result["context"]
+    assert "cpuarch" in result["context"]
+    assert "python_version" in result["context"]
+
+
+def test_error_on_duplicate_context_keys_in_runner(testfolder: str) -> None:
+    r = BenchmarkRunner()
+
+    def duplicate_context_provider() -> dict[str, str]:
+        return {"system": "DuplicateSystem"}
+
+    context_providers = [system, duplicate_context_provider]
+
+    with pytest.raises(ValueError) as e:
+        r.run(
+            os.path.join(testfolder),
+            tags=("standard",),
+            params={"x": 1, "y": 1},
+            context=context_providers,
+        )
+
+    assert "got multiple values for context key 'system'" in str(e.value)
```

Comment:
```suggestion
    with pytest.raises(ValueError, match="got multiple values for context key 'system'"):
        r.run(
            testfolder,
            tags=("standard",),
            params={"x": 1, "y": 1},
            context=context_providers,
        )
```

Any particular reason you're calling `os.path.join()` on a single directory all the time? For the trailing slash or?

Reply:
No, that was hwo the existing tests were implemented and I rolled with it. It's not needed for single dirs or files. Removing.